udmis/src/main/java/com/google/bos/udmi/monitoring/LogTail.java

```diff
@@ -0,0 +1,253 @@
+import com.google.cloud.audit.AuditLog;
+import com.google.cloud.logging.LogEntry;
+import com.google.cloud.logging.LogEntryServerStream;
+import com.google.cloud.logging.Logging;
+import com.google.cloud.logging.Logging.TailOption;
+import com.google.cloud.logging.LoggingOptions;
+import com.google.cloud.logging.Payload;
+import com.google.cloud.logging.Severity;
+import com.google.protobuf.InvalidProtocolBufferException;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.time.Instant;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import udmi.schema.Monitoring;
+import udmi.schema.MonitoringMetric;
+
+public class LogTail {
+
+  private String projectId;
+  private LogTimeSeries logsTimeSeries;
+  private String LOG_FILTER =
+      "(resource.labels.function_name=\"udmi_target\") OR " +
+          "(resource.labels.function_name=\"udmi_state\") OR " +
+          "(resource.labels.function_name=\"udmi_config\") OR " +
+          "(severity=ERROR AND protoPayload.serviceName=\"cloudiot.googleapis.com\")";
+
+  public LogTail(String projectId) {
+    this.projectId = projectId;
+    this.logsTimeSeries = new LogTimeSeries();
+  }
+
+  public static void main(String[] args) {
+    String projectId = "essential-keep-197822";
+    LogTail lt = new LogTail(projectId);
+    lt.tailLogs();
+  }
+
+  private LogEntryServerStream getCloudLogStream(String filter) {
+    LoggingOptions options = LoggingOptions.getDefaultInstance();
+    Logging logging = options.getService();
+    LogEntryServerStream stream;
+    if (filter == null) {
+      stream = logging.tailLogEntries();
+    } else {
+      stream = logging.tailLogEntries(TailOption.filter(filter));
+    }
+    return stream;
+  }
+
+  /*
+   * Top level processor of incoming GCP log entries.
+   */
+  private void processLogEntry(LogEntry log) {
+    if (log.getSeverity().equals(Severity.ERROR)) {
+      processLogEntryError(log);
+    } else {
+      String function_name = log.getResource().getLabels().getOrDefault("function_name", null);
+      if (function_name.equals("udmi_target")) {
+      } else if (function_name.equals("udmi_state")) {
+      } else if (function_name.equals("udmi_config")) {
+      }
+    }
+  }
+
+  /*
+   * Top level processor of incoming GCP log entries of ERROR type.
+   */
+  private void processLogEntryError(LogEntry log) {
+    // Be careful in here. Unwrapping this is a careful sequence.
+    Payload.ProtoPayload payload = log.getPayload();
+    if (!payload.getData().getTypeUrl().equals("type.googleapis.com/google.cloud.audit.AuditLog")) {
+      System.err.println("Not AuditLog");
+      return;
+    }
+    try {
+      AuditLog auditLog = AuditLog.parseFrom(payload.getData().getValue().toByteArray());
+      LogTailEntry entry = new LogTailEntry();
+      entry.timestamp = log.getInstantTimestamp();
```

Comment:
field needs to be converted to an iso timestamp -- there's some utility functions in GeneralUtils or JsonUtils that do this (look for "timestamp" or something like that).   Or an Instant might have the right thing, I forget.

Reply:
Missed this. Will address.

Reply:
Just checking, you mean at the output stage, right? The JSON stringify already wrote in something like ISO8601.